C3 Qemu support

mzakharo opened the pull request:
<!-- Thanks for submitting a Pull Request! We appreciate you spending the
     time to improve MicroPython. Please provide enough information so that
     others can review your Pull Request.

     Before submitting, please read:
     https://github.com/micropython/micropython/blob/master/CODEOFCONDUCT.md
     https://github.com/micropython/micropython/wiki/ContributorGuidelines

     Please check any CI failures that appear after your Pull Request is opened.
-->

### Summary

Add support for C3 target to run on QEMU emulator with LAN networking support


### Testing

Tested with IDF 5.3.2

`make BOARD=ESP32_GENERIC_C3 BOARD_VARIANT=QEMU qemu`
```python
>>> import network
>>> lan = network.LAN(phy_type=network.PHY_OPENETH, phy_addr=1)
>>> lan.active(True)
>>> lan.ipconfig('addr4')
('10.0.2.15', '255.255.255.0')
```


### Trade-offs and Alternatives
`esp_efuse_rtc_calib_get_ver()`, part of `efuse` component, does not play nicely with the emulator.  Presented here is an example workaround for the issue, outlined here: https://github.com/espressif/esp-insights/issues/38

dpgeorge commented on ports/esp32/boards/ESP32_QEMU_C3/board.md:
```diff
@@ -0,0 +1,7 @@
+
+USB serial/JTAG support is enabled on pin 18 and 19. Note that this
+is not a full USB stack, the C3 just provides a CDC/ACM class serial
+and JTAG interface.
```
So, this board also works on hardware as well as qemu?  Won't the ETH break on hardware?

mzakharo replied:
We can make it a variant of esp32 generic c3 instead.

mzakharo replied:
OPENETH usage will break on hardware, other than that the build is compatible with both.

projectgus replied:
I've marked as un-resolved to keep track of this:  either this should be a board variant, or this text should be changed to describe what this "board" is rather than being copy-paste from the C3 board.

mzakharo replied:
Changed to board variant

dpgeorge commented on ports/esp32/network_lan.c:
```diff
@@ -232,6 +235,13 @@ static mp_obj_t get_lan(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_ar
             self->phy = esp_eth_phy_new_ksz80xx(&phy_config);
             break;
         #endif
+        #if CONFIG_ETH_USE_OPENETH
```
Please split this addition of `PYH_OPENETH` into a separate commit within this PR.

mzakharo replied:
Will do

dpgeorge commented on ports/esp32/main.c:
```diff
@@ -257,3 +257,15 @@ void *esp_native_code_commit(void *buf, size_t len, void *reloc) {
+extern int  __real_esp_efuse_rtc_calib_get_ver(void);
+int __wrap_esp_efuse_rtc_calib_get_ver(void)
+{
+    #if MICROPY_PI_QEMU
```
I don't know if this is the correct name for such a config option.

Is there a macro option that the IDF defines when building for qemu that can be used here instead?

mzakharo replied:
The idf build has nothing special when targeting qemu. Enabling openeth is the only way to tell it's a qemu build

mzakharo replied:
switched it to CONFIG_ETH_USE_OPENETH instead and removed custom config option

dpgeorge commented on ports/esp32/Makefile:
```diff
@@ -106,6 +106,12 @@ size-components:
+	$(call RUN_IDF_PY,qemu monitor)
+
+menuconfig:
+	$(call RUN_IDF_PY,menuconfig)
```
Is this `menuconfig` target needed?

mzakharo replied:
Will remove

dpgeorge reviewed (commented):
Thanks, this looks interesting!  I did a very quick review so maybe more changes are needed.

mzakharo committed:
esp32/network_lan: Add support for OPENETH LAN PHY.

Can be used with QEMU.

Signed-off-by: Mikhail Zakharov <mzakharo@gmail.com>

projectgus commented on ports/esp32/boards/ESP32_QEMU_C3/board.json:
```diff
@@ -0,0 +1,19 @@
+    "mcu": "esp32c3",
+    "product": "ESP32-C3",
+    "thumbnail": "",
+    "url": "https://www.espressif.com/en/products/modules",
```
Suggest changing some of these to reflect what this actually is,

i.e. `url` can link to https://docs.espressif.com/projects/esp-idf/en/latest/esp32c3/api-guides/tools/qemu.html, `product` can be "ESP32-C3 QEMU Emulator", etc.

mzakharo replied:
changed to board variant instead

mzakharo committed:
esp32: Add QEMU support for ESP32_GENERIC_C3 board.

Add support for C3 target to run on QEMU emulator with LAN networking
support.

Build and run via:

    make BOARD=ESP32_GENERIC_C3 BOARD_VARIANT=QEMU qemu

Signed-off-by: Mikhail Zakharov <mzakharo@gmail.com>

dpgeorge reviewed (approved):
Thanks for updating, and sorry this took ages to get back to.

It looks good now, nice and simple as a board variant.

I tested on IDF 5.5.2 and it works well (after installing their custom qemu fork).
